### PR TITLE
Audio: add sound event detection inference stage

### PR DESCRIPTION
**Pull request by mohammadaaftabv**
## What this does

Adds a generic `SEDInferenceStage` that runs sound-event detection over audio and emits framewise class probabilities.

This PR contains only the SED producer. Event thresholding, interval construction, filtering, and event-to-task fan-out remain in #2261.

## Stage / adapter split

The implementation follows the same boundary as the generic ASR stage:

- `nemo_curator/stages/audio/inference/base.py` owns behavior shared by adapter-backed audio stages: lazy Hydra adapter resolution, resource-to-physical-GPU conversion, load-failure cleanup, worker teardown, file-versus-waveform input declaration, and SoundFile decoding.
- `nemo_curator/stages/audio/inference/asr/stage.py` and `nemo_curator/stages/audio/inference/sed.py` inherit that shared infrastructure and retain only task/model-specific preparation, inference, and output assembly.
- Worker sizing uses the existing `ProcessingStage.with_(num_workers=...)` mechanism. SED no longer carries a separate `num_workers_override` field or `num_workers()` implementation.
- `nemo_curator/models/sed/base.py` defines the canonical `SEDAdapter` / `SEDResult` contract.
- `nemo_curator/models/sed/panns.py` owns PANNs checkpoint loading, device selection, model-specific padding, one batched inference call, and temporal result metadata.
- `adapter_target` remains lazy and YAML-selectable, so another runtime can replace PANNs without changing the stage's task schema.
- CNN14 and its registry live under `nemo_curator/models/sed/`; no model implementation remains under the stages package.

## Runnable YAML

A complete example is included at `tutorials/audio/sed/pipeline.yaml`. Each input JSONL row contains an audio path:

```json
{"audio_filepath": "/absolute/path/to/audio.wav"}
```

Run from the Curator repository root:

```bash
uv run --extra audio_sed python nemo_curator/config/run.py \
  --config-path ../../tutorials/audio/sed \
  --config-name pipeline \
  manifest_path=/absolute/path/to/input.jsonl \
  'checkpoint_path=/absolute/path/to/Cnn14_DecisionLevelMax_mAP\=0.385.pth' \
  output_dir=/absolute/path/to/sed_output
```

The checkpoint value is quoted and its `=` is escaped for Hydra. Use the published decision-level SED checkpoint, [`Cnn14_DecisionLevelMax_mAP=0.385.pth`](https://zenodo.org/record/3987831/files/Cnn14_DecisionLevelMax_mAP%3D0.385.pth?download=1). The clip-level `Cnn14_mAP=0.431.pth` is not compatible with the configured decision-level architecture.

The YAML uses the checkpoint-native 32 kHz frontend and writes persistent compressed NPZ sidecars. The same command and input contract are documented in the stage module docstring; no additional README is introduced.

## Output contract

| Key | Meaning |
| --- | --- |
| `_sed_framewise` | `[frame, class]` probability matrix |
| `sed_fps` | Adapter-reported frame rate |
| `sed_valid_frames` | Number of leading frames belonging to real audio rather than model/batch padding |
| `npz_filepath` | Optional deterministic compressed sidecar path |

NPZ names use Curator's deterministic path hash, so same-named source files do not collide and reruns produce stable paths.

## Model provenance and dependency

The three PANNs CNN14 decision-level variants live in `nemo_curator/models/sed/cnn14.py`.

The implementation is derived from [qiuqiangkong/audioset_tagging_cnn](https://github.com/qiuqiangkong/audioset_tagging_cnn), which is MIT licensed, copyright 2018–2020 Qiuqiang Kong. The complete MIT notice and source attribution remain at the top of the model file.

The registry and stage remain lazy: importing them does not import `torchlibrosa`. Constructing a PANNs model requires the opt-in `audio_sed` extra.

## Current diff

- 16 files
- +1991 / -68
- base: `main@0d94617b93b7d58249a4e0ac6803d1851eaaec81`
- head: `89070f8b2e4382ccf396892f199dfc71679e0d62`

## Validation

Current head:

- Ruff passed for all changed shared-base, ASR, SED, and test paths.
- `git diff --check` and Python byte-compilation passed.
- Shared base plus ASR/SED stage suites: 90 passed.
- PANNs adapter and CNN14 model suites: 27 passed.
- The example YAML was run locally through Ray Data on an NVIDIA RTX 3080 Ti using the two-entry audio-tagging fixture manifest after the shared-base refactor.
- The PANNs adapter loaded `Cnn14_DecisionLevelMax` on CUDA and processed both waveforms in one CNN14 call; the pipeline completed with two final tasks.
- Exactly two NPZ sidecars were produced. Both are `float16`, have shape `(6708, 527)`, contain only finite probabilities in `[0, 1]`, and report 100 fps.
- Duration reconstruction was exact for the 60.0 s sample and within 6.5 ms for the 67.0735 s sample.
- Tested checkpoint SHA-256: `dd3b4043a87d4ec13df8082c0fcfee3fb5084151808e47e060987a95eabdd142`.

Earlier producer-stage numerical parity was run at pre-refactor head `1328caa321fb723c489bd69c102c00e9eb305ff0` against `nithinraok/Curator:nkoluguri/integration-test@2963bf3a23d807ff26a80cf80aeb1a33fd0f7172`: both two-row archives were byte-identical with SHA-256 `3abb4430b1d0568ed995199d216394c65d5783074f67a0a64eaca4e351823ae6`, and maximum/mean framewise deltas were `0.0`. That numerical run predates the stage/adapter-only refactor in the current head.

### CODE REVIEW

File: nemo_curator/models/sed/cnn14.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""CNN14 decision-level model implementations for Sound Event Detection.
+
+Vendored from the PANNs reference implementation at
+https://github.com/qiuqiangkong/audioset_tagging_cnn (MIT, copyright
```
Comment (sarahyurick) [inline nemo_curator/models/sed/cnn14.py]: Is this file copy-pasted from a more specific link?

File: nemo_curator/models/sed/panns.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    mel_bins: int = 64
+    fmin: int = 50
+    fmax: int = 14000
+    classes_num: int = 527
+    pad_short_segments: bool = True
+    _model: Any = field(default=None, init=False, repr=False)
+    _device: Any = field(default=None, init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        if not self.checkpoint_path:
```
Comment (sarahyurick) [inline nemo_curator/models/sed/panns.py]: This block will never fire right?

File: nemo_curator/models/sed/panns.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        """Load checkpoint weights on CPU first, then place the model."""
+        if self._model is not None:
+            return
+        if not isinstance(num_gpus, int) or isinstance(num_gpus, bool) or num_gpus not in {0, 1}:
+            msg = f"PANNsSEDAdapter supports zero or one physical GPU, got {num_gpus!r}"
+            raise ValueError(msg)
+
+        import torch
+
+        from nemo_curator.models.sed import get_model_class
```
Comment (sarahyurick) [inline nemo_curator/models/sed/panns.py]: Can these be top-level imports?

File: nemo_curator/stages/audio/inference/sed/__init__.py
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/__init__.py:1]: No copyright needed for an empty file.

File: nemo_curator/stages/audio/inference/sed/stage.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    framewise_dtype: str = "float16"
+    skip_if_output_exists: bool = False
+
+    adapter_kwargs: dict[str, Any] = field(default_factory=dict)
+    batch_size: int = 32
+    resources: Resources = field(default_factory=lambda: Resources(cpus=1.0, gpu_memory_gb=4.0))
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        if not self.checkpoint_path:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/stage.py]: Similar comment as above.

File: nemo_curator/stages/audio/inference/sed/stage.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            msg = "SEDInferenceStage.checkpoint_path must be non-empty"
+            raise ValueError(msg)
+        if not isinstance(self.sample_rate, int) or isinstance(self.sample_rate, bool) or self.sample_rate <= 0:
+            msg = f"SEDInferenceStage.sample_rate must be a positive integer, got {self.sample_rate!r}"
+            raise ValueError(msg)
+        if not isinstance(self.batch_size, int) or isinstance(self.batch_size, bool) or self.batch_size <= 0:
+            msg = f"SEDInferenceStage.batch_size must be a positive integer, got {self.batch_size!r}"
+            raise ValueError(msg)
+        if self.framewise_dtype not in {"float16", "float32"}:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/stage.py:131]: Can the typehint be updated to reflect this?

File: nemo_curator/stages/audio/inference/sed/stage.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        output_keys = [self.framewise_output_key, self.valid_frames_key, self.fps_key]
+        if self.save_npz:
+            output_keys.append(self.npz_filepath_key)
+        return [], output_keys
+
+    def _prepare_waveform(self, waveform: object, sample_rate: object) -> np.ndarray:
+        """Return contiguous mono float32 samples at the adapter sample rate."""
+        from nemo_curator.stages.audio.common import ensure_mono, ensure_waveform_2d
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/stage.py]: Similar comment as above.

File: nemo_curator/stages/audio/inference/sed/stage.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        if source_sample_rate != self.sample_rate:
+            import librosa
+
+            prepared = librosa.resample(prepared, orig_sr=source_sample_rate, target_sr=self.sample_rate)
+            prepared = np.ascontiguousarray(prepared, dtype=np.float32)
+        return prepared
+
+    def process(self, task: AudioTask) -> AudioTask:
+        """Run one task through the same batch path used by executors."""
+        return self.process_batch([task])[0]
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/stage.py]: It should never reach this is `process_batch` is implemented by the stage right? So maybe it should be a `NotImplementedError`.

File: nemo_curator/stages/audio/inference/base.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    adapter_target: str
+    waveform_key: str | None
+    sample_rate_key: str
+    audio_filepath_key: str
+    resources: Resources
+    _adapter: AdapterT | None
+
+    def __post_init__(self) -> None:
+        if not self.adapter_target:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/base.py]: Similar comment as above.

File: tests/models/sed/__init__.py
Comment (sarahyurick) [inline tests/models/sed/__init__.py:1]: This can be empty.

File: pyproject.toml [hunk 9]
```diff
@@ -141,6 +141,12 @@ audio_cuda12 = [
     "torchcodec; platform_machine == 'x86_64' and platform_system != 'Darwin'",
 ]
 
+# Sound event detection; opt-in because torchlibrosa is only used by the CNN14
+# SED front-end and nothing else in the audio stack needs it.
+audio_sed = [
+    "torchlibrosa>=0.1.0",
```
Comment (sarahyurick) [inline pyproject.toml:141]: I see we already have `librosa` in `audio_common`, should we just do `torchlibrosa` in `audio_cuda12`?

File: nemo_curator/models/sed/panns.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        return torch.hub.load_state_dict_from_url(
+            _DEFAULT_CHECKPOINT_URL,
+            map_location="cpu",
+            progress=False,
+            file_name=_DEFAULT_CHECKPOINT_FILENAME,
+            weights_only=True,
+        )
+
+    def download_weights_on_node(self) -> None:
+        """Populate PyTorch Hub's cache for the registered default checkpoint."""
```
Comment (sarahyurick) [inline nemo_curator/models/sed/panns.py]: What if I want the model to live somewhere other than the HF cache?

File: nemo_curator/models/sed/panns.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+            self._load_checkpoint()
+
+    def load_model(self, *, num_gpus: int) -> None:
+        """Load checkpoint weights on CPU first, then place the model."""
+        if not isinstance(num_gpus, int) or isinstance(num_gpus, bool) or num_gpus not in {0, 1}:
+            msg = f"PANNsSEDAdapter supports zero or one physical GPU, got {num_gpus!r}"
+            raise ValueError(msg)
+
+        model_cls = get_model_class(self.model_type)
+        use_cuda = num_gpus == 1 and torch.cuda.is_available()
```
Comment (sarahyurick) [inline nemo_curator/models/sed/panns.py]: What if `num_gpus` > 1?

File: nemo_curator/stages/audio/inference/sed/stage.py [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+
+        results = cast("SEDAdapter", self._adapter).infer_batch(items)
+        if len(results) != len(items):
+            msg = f"SED adapter returned {len(results)} results for {len(items)} items (must match 1:1)"
+            raise RuntimeError(msg)
+        self._write_results(tasks, valid_indices, results, audio_paths)
+
+        skipped_count = len(tasks) - len(valid_indices)
+        if skipped_count:
+            logger.info("SED batch: skipped {}/{} tasks", skipped_count, len(tasks))
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/inference/sed/stage.py]: Should this be a warning?

File: tests/models/sed/test_cnn14.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    pytest.importorskip("torchlibrosa")
+    from torch import nn
+
+    assert issubclass(sed.get_model_class(model_type), nn.Module)
+
+
+@pytest.mark.parametrize("model_type", ["Cnn14_DecisionLevelMax", "Cnn14_DecisionLevelAvg", "Cnn14_DecisionLevelAtt"])
+def test_each_variant_emits_framewise_and_clipwise_output(model_type: str) -> None:
+    """SED needs framewise output; the base Cnn14 gives clip-level only."""
+    pytest.importorskip("torchlibrosa")
```
Comment (sarahyurick) [inline tests/models/sed/test_cnn14.py]: Let's remove the `importorskip`, it is in the pyproject so it should work.

File: tests/models/sed/test_panns.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+"""Tests for the PANNs implementation of the SED adapter contract."""
+
+from pathlib import Path
+from unittest.mock import MagicMock, patch
+
+import numpy as np
+import pytest
+
+torch = pytest.importorskip("torch")
```
Comment (sarahyurick) [inline tests/models/sed/test_panns.py]: Same here, let's remove `importorskip`.

**Review by virajkarandikar — APPROVED** [review]

File: nemo_curator/models/sed/panns.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+            map_location="cpu",
+            progress=False,
+            file_name=_DEFAULT_CHECKPOINT_FILENAME,
+            weights_only=True,
+        )
+
+    def download_weights_on_node(self) -> None:
+        """Populate the configured PyTorch Hub cache for the default checkpoint."""
+        if self.checkpoint_path is None:
+            self._load_checkpoint()
```
Comment (sarahyurick) [inline nemo_curator/models/sed/panns.py]: Left a previous comment before. Shouldn't `checkpoint_path` be determining where to download and/or load the model from?

**Review by sarahyurick — APPROVED** [review]
Thanks!

**Comment by mohammadaaftabv** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Curator/pull/2260/commits/41ad47a2ab31068a6a05fb26398193064d8ef60d
